File: examples/typescript/ws/function/shared/protocol.ts
// Message type enums
export enum ClientMessageType {
  SEND = "send",
  DISCONNECT = "disconnect",
}

export enum ServerMessageType {
  BROADCAST = "broadcast",
  CONNECTED = "connected",
  USER_JOINED = "user_joined",
  USER_LEFT = "user_left",
  ERROR = "error",
  ACK = "ack",
}

// Client → Server messages
export type SendMessage = { type: ClientMessageType.SEND; message: string };
export type DisconnectMessage = { type: ClientMessageType.DISCONNECT };
export type ClientMessage = SendMessage | DisconnectMessage;

// Helper functions for creating client messages
export function createSendMessage(message: string): SendMessage {
  return { type: ClientMessageType.SEND, message };
}

export function createDisconnectMessage(): DisconnectMessage {
  return { type: ClientMessageType.DISCONNECT };
}

// Server → Client messages
export type BroadcastMessage = {
  type: ServerMessageType.BROADCAST;
  from: string;
  message: string;
};
export type ConnectedMessage = { type: ServerMessageType.CONNECTED; userId: string };
export type UserJoinedMessage = { type: ServerMessageType.USER_JOINED; userId: string };
export type UserLeftMessage = { type: ServerMessageType.USER_LEFT; userId: string };
export type ErrorMessage = { type: ServerMessageType.ERROR; message: string };
export type AckMessage = { type: ServerMessageType.ACK; message: string };
export type ServerMessage =
  | BroadcastMessage
  | ConnectedMessage
  | UserJoinedMessage
  | UserLeftMessage
  | ErrorMessage
  | AckMessage;

// Helper functions for creating server messages
export function createBroadcastMessage(from: string, message: string): BroadcastMessage {
  return { type: ServerMessageType.BROADCAST, from, message };
}

export function createConnectedMessage(userId: string): ConnectedMessage {
  return { type: ServerMessageType.CONNECTED, userId };
}

export function createUserJoinedMessage(userId: string): UserJoinedMessage {
  return { type: ServerMessageType.USER_JOINED, userId };
}

export function createUserLeftMessage(userId: string): UserLeftMessage {
  return { type: ServerMessageType.USER_LEFT, userId };
}

export function createErrorMessage(message: string): ErrorMessage {
  return { type: ServerMessageType.ERROR, message };
}

export function createAckMessage(message: string): AckMessage {
  return { type: ServerMessageType.ACK, message };
}

// Helper to check if an object has a type property
function hasType(msg: unknown): msg is { type: unknown } {
  return typeof msg === "object" && msg !== null && "type" in msg;
}

export function isSendMessage(msg: unknown): msg is SendMessage {
  return hasType(msg) && msg.type === ClientMessageType.SEND;
}

export function isDisconnectMessage(msg: unknown): msg is DisconnectMessage {
  return hasType(msg) && msg.type === ClientMessageType.DISCONNECT;
}

// Server message type guards
export function isBroadcastMessage(msg: unknown): msg is BroadcastMessage {
  return hasType(msg) && msg.type === ServerMessageType.BROADCAST;
}

export function isConnectedMessage(msg: unknown): msg is ConnectedMessage {
  return hasType(msg) && msg.type === ServerMessageType.CONNECTED;
}

export function isUserJoinedMessage(msg: unknown): msg is UserJoinedMessage {
  return hasType(msg) && msg.type === ServerMessageType.USER_JOINED;
}

export function isUserLeftMessage(msg: unknown): msg is UserLeftMessage {
  return hasType(msg) && msg.type === ServerMessageType.USER_LEFT;
}

export function isErrorMessage(msg: unknown): msg is ErrorMessage {
  return hasType(msg) && msg.type === ServerMessageType.ERROR;
}

export function isAckMessage(msg: unknown): msg is AckMessage {
  return hasType(msg) && msg.type === ServerMessageType.ACK;
}

// Parse and validate client message from JSON string
export function parseClientMessage(body: string): ClientMessage {
  // Parse JSON
  let parsed: unknown;
  try {
    parsed = JSON.parse(body);
  } catch (error) {
    throw new Error("Invalid JSON");
  }

  // Validate that parsed message has a type field
  if (!parsed || typeof parsed !== "object" || !("type" in parsed)) {
    throw new Error("Invalid message format");
  }

  // Validate message type
  const messageType = (parsed as { type: unknown }).type;
  if (
    messageType !== ClientMessageType.SEND &&
    messageType !== ClientMessageType.DISCONNECT
  ) {
    throw new Error("Unknown message type");
  }

  return parsed as ClientMessage;
}

// Data Streams Trigger → Handler messages
export enum TriggerMessageType {
  DATA_STREAMS = "data_streams"
}

// According to broadcast.md, trigger sends: { "messages": [{...}, {...}] }
// We'll send all server messages (BROADCAST, USER_JOINED, USER_LEFT) via topic
export type DataStreamsTriggerMessage = {
  messages: ServerMessage[]  // Array of any server messages to broadcast
}

export type TriggerMessage = DataStreamsTriggerMessage

// Helper to create trigger message
export function createDataStreamsTriggerMessage(
  messages: ServerMessage[]
): DataStreamsTriggerMessage {
  return { messages }
}

// Type guard
export function isDataStreamsTriggerMessage(msg: unknown): msg is DataStreamsTriggerMessage {
  return (
    typeof msg === "object" &&
    msg !== null &&
    "messages" in msg &&
    Array.isArray((msg as any).messages)
  )
}

File: examples/go/ws/function/server/protocol.go
package main

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"time"
)

// Message creation functions

// CreateConnectedMessage creates a CONNECTED message for the client
func CreateConnectedMessage(userID string) ServerMessage {
	return ServerMessage{
		Type:      MessageTypeConnected,
		UserID:    userID,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
}

// CreateBroadcastMessage creates a BROADCAST message for all clients
func CreateBroadcastMessage(userID, content string) ServerMessage {
	return ServerMessage{
		Type:      MessageTypeBroadcast,
		UserID:    userID,
		Content:   content,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
}

// CreateUserJoinedMessage creates a USER_JOINED message
func CreateUserJoinedMessage(userID string) ServerMessage {
	return ServerMessage{
		Type:      MessageTypeUserJoined,
		UserID:    userID,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
}

// CreateUserLeftMessage creates a USER_LEFT message
func CreateUserLeftMessage(userID string) ServerMessage {
	return ServerMessage{
		Type:      MessageTypeUserLeft,
		UserID:    userID,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
}

// CreateErrorMessage creates an ERROR message
func CreateErrorMessage(message, code string) ServerMessage {
	return ServerMessage{
		Type:      MessageTypeError,
		Message:   message,
		Code:      code,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
	}
}

// CreateAckMessage creates an ACK message
func CreateAckMessage(originalType string) ServerMessage {
	return ServerMessage{
		Type:         MessageTypeAck,
		OriginalType: originalType,
		Timestamp:    time.Now().UTC().Format(time.RFC3339),
	}
}

// ParseClientMessage parses and validates a client message from JSON
func ParseClientMessage(data []byte) (*ClientMessage, error) {
	var msg ClientMessage
	if err := json.Unmarshal(data, &msg); err != nil {
		return nil, fmt.Errorf("failed to parse client message: %w", err)
	}

	// Validate message type
	if msg.Type != "SEND" && msg.Type != "DISCONNECT" {
		return nil, fmt.Errorf("invalid message type: %s", msg.Type)
	}

	// Validate SEND messages have content
	if msg.Type == "SEND" && msg.Content == "" {
		return nil, fmt.Errorf("SEND message must have content")
	}

	return &msg, nil
}

// DecodeMessageBody decodes the message body, handling base64 encoding if necessary
func DecodeMessageBody(body string, isBase64Encoded bool) ([]byte, error) {
	if isBase64Encoded {
		decoded, err := base64.StdEncoding.DecodeString(body)
		if err != nil {
			return nil, fmt.Errorf("failed to decode base64 body: %w", err)
		}
		return decoded, nil
	}
	return []byte(body), nil
}

// SerializeMessage serializes a ServerMessage to JSON
func SerializeMessage(msg ServerMessage) ([]byte, error) {
	data, err := json.Marshal(msg)
	if err != nil {
		return nil, fmt.Errorf("failed to serialize message: %w", err)
	}
	return data, nil
}

// SerializeTriggerMessage serializes a TriggerMessage (wrapper with messages array)
func SerializeTriggerMessage(messages []ServerMessage) ([]byte, error) {
	wrapper := TriggerMessage{
		Messages: messages,
	}
	data, err := json.Marshal(wrapper)
	if err != nil {
		return nil, fmt.Errorf("failed to serialize trigger message: %w", err)
	}
	return data, nil
}
